### PR TITLE
Spark: Add connector APIs for DELETE/UPDATE/MERGE

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR adds APIs to support DELETE/UPDATE/MERGE operations in Spark.

### CODE REVIEW

**Comment by aokolnychyi** [comment]
These interfaced are added to the iceberg module too.

File: spark3/src/main/java/org/apache/spark/sql/connector/iceberg/catalog/ExtendedSupportsDelete.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * under the License.
+ */
+
+package org.apache.spark.sql.connector.iceberg.catalog;
+
+import org.apache.spark.sql.connector.catalog.SupportsDelete;
+import org.apache.spark.sql.sources.Filter;
+
+// this should be part of SupportsDelete when merged upstream
+public interface ExtendedSupportsDelete extends SupportsDelete {
```
Comment (aokolnychyi) [inline spark3/src/main/java/org/apache/spark/sql/connector/iceberg/catalog/ExtendedSupportsDelete.java:26]: Seems like Spark is about to freeze 3.1 from Monday so it is unfortunate we did not get that into Spark :(

Comment (aokolnychyi) [reply spark3/src/main/java/org/apache/spark/sql/connector/iceberg/catalog/ExtendedSupportsDelete.java:26]: I'll submit a PR next week anyway.

Comment (rdblue) [reply spark3/src/main/java/org/apache/spark/sql/connector/iceberg/catalog/ExtendedSupportsDelete.java:26]: We can get it in still. This is a small change.

Comment (rdblue) [reply spark3/src/main/java/org/apache/spark/sql/connector/iceberg/catalog/ExtendedSupportsDelete.java:26]: And the freeze date under discussion was at least 1 December, so we have some time.

File: spark3/src/main/java/org/apache/spark/sql/connector/iceberg/catalog/SupportsMerge.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ * under the License.
+ */
+
+package org.apache.spark.sql.connector.iceberg.catalog;
+
+import org.apache.spark.sql.connector.catalog.Table;
+import org.apache.spark.sql.connector.iceberg.write.MergeBuilder;
+import org.apache.spark.sql.connector.write.LogicalWriteInfo;
+
+public interface SupportsMerge extends Table {
```
Comment (rdblue) [inline spark3/src/main/java/org/apache/spark/sql/connector/iceberg/catalog/SupportsMerge.java:26]: Can we add Javadoc for all of these classes and methods? I think that would be helpful.

Comment (aokolnychyi) [reply spark3/src/main/java/org/apache/spark/sql/connector/iceberg/catalog/SupportsMerge.java:26]: Yeah, let me spend some time to add docs to the existing classes too. I think we are relatively confident in the implementation now.

File: spark3/src/main/java/org/apache/spark/sql/connector/iceberg/catalog/SupportsMerge.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ */
+
+package org.apache.spark.sql.connector.iceberg.catalog;
+
+import org.apache.spark.sql.connector.catalog.Table;
+import org.apache.spark.sql.connector.iceberg.write.MergeBuilder;
+import org.apache.spark.sql.connector.write.LogicalWriteInfo;
+
+/**
+ * A mix-in interface for Table to indicate that it supports row-level operations.
```
Comment (aokolnychyi) [inline spark3/src/main/java/org/apache/spark/sql/connector/iceberg/catalog/SupportsMerge.java:27]: I did not reference Spark classes in our Javadoc.

File: spark3/src/main/java/org/apache/spark/sql/connector/iceberg/catalog/SupportsMerge.java (hunk 2 above)
Comment (aokolnychyi) [reply spark3/src/main/java/org/apache/spark/sql/connector/iceberg/catalog/SupportsMerge.java:32]: I added some basic docs. Feel free to correct wording if it is not clear.

**Review by rdblue — APPROVED** [review]

**Pull request merged**